crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheMetadata.java

```diff
@@ -29,12 +33,87 @@
 
   private final ImageLayers<CachedLayerWithMetadata> layers = new ImageLayers<>();
 
+  /** Can be used to filter layers in the metadata. */
+  static class LayerFilter {
+
+    private final ImageLayers<CachedLayerWithMetadata> layers;
+
+    @Nullable private CachedLayerType type;
+    @Nullable private Set<File> sourceFiles;
+
+    /** True if the filters are used; false otherwise. */
+    private boolean isTypeFilterEnabled = false;
+
+    private boolean isSourceFilesFilterEnabled = false;
+
+    private LayerFilter(ImageLayers<CachedLayerWithMetadata> layers) {
+      this.layers = layers;
+    }
+
+    /** Filters to a certain layer type. */
+    LayerFilter byType(CachedLayerType type) {
+      this.type = type;
```

Comment:
This is weird. I'd expect a method like this to return a new object, not modify the existing object. Maybe it's just the method name that's throwing me.

Reply:
What if someone passes null to these methods? That still sets isTypeFilterEnabled to true.

Reply:
Removed isTypeFilterEnabled.